packages/core/src/utils/request.ts

```diff
@@ -0,0 +1,135 @@
+import type { PolymorphicRequest, RequestEventData } from '../types-hoist';
+import type { WebFetchHeaders, WebFetchRequest } from '../types-hoist/webfetchapi';
+import { dropUndefinedKeys } from '../utils-hoist/object';
+
+/**
+ * Transforms a `Headers` object that implements the `Web Fetch API` (https://developer.mozilla.org/en-US/docs/Web/API/Headers) into a simple key-value dict.
+ * The header keys will be lower case: e.g. A "Content-Type" header will be stored as "content-type".
+ */
+export function winterCGHeadersToDict(winterCGHeaders: WebFetchHeaders): Record<string, string> {
+  const headers: Record<string, string> = {};
+  try {
+    winterCGHeaders.forEach((value, key) => {
+      if (typeof value === 'string') {
+        // We check that value is a string even though it might be redundant to make sure prototype pollution is not possible.
+        headers[key] = value;
+      }
+    });
+  } catch {
+    // just return the empty headers
+  }
+
+  return headers;
+}
+
+/**
+ * Convert common request headers to a simple dictionary.
+ */
+export function headersToDict(reqHeaders: Record<string, string | string[] | undefined>): Record<string, string> {
+  const headers: Record<string, string> = Object.create(null);
+
+  try {
+    Object.entries(reqHeaders).forEach(([key, value]) => {
+      if (typeof value === 'string') {
+        headers[key] = value;
+      }
+    });
+  } catch {
+    // just return the empty headers
+  }
+
+  return headers;
+}
+
+/**
+ * Converts a `Request` object that implements the `Web Fetch API` (https://developer.mozilla.org/en-US/docs/Web/API/Headers) into the format that the `RequestData` integration understands.
+ */
+export function winterCGRequestToRequestData(req: WebFetchRequest): RequestEventData {
+  const headers = winterCGHeadersToDict(req.headers);
+
+  return {
+    method: req.method,
+    url: req.url,
+    query_string: extractQueryParamsFromUrl(req.url),
+    headers,
+    // TODO: Can we extract body data from the request?
+  };
+}
+
+/**
+ * Convert a HTTP request object to RequestEventData to be passed as normalizedRequest.
+ * Instead of allowing `PolymorphicRequest` to be passed,
+ * we want to be more specific and generally require a http.IncomingMessage-like object.
+ */
+export function httpRequestToRequestData(request: {
+  method?: string;
+  url?: string;
+  headers?: {
+    [key: string]: string | string[] | undefined;
+  };
+  protocol?: string;
+  socket?: {
+    encrypted?: boolean;
+    remoteAddress?: string;
+  };
+}): RequestEventData {
+  const headers = request.headers || {};
+  const host = typeof headers.host === 'string' ? headers.host : undefined;
+  const protocol = request.protocol || (request.socket?.encrypted ? 'https' : 'http');
+  const url = request.url || '';
+
+  const absoluteUrl = getAbsoluteUrl({
```

Comment:
previously, we always built this from the host. However, the host could also be `<no host>` as fallback, leading to (theoretically) urls like: `http://<no host>/` which is weird. Now, we simply have no URL in such a case.